cli/src/commands/run.rs

```diff
@@ -14,12 +14,262 @@
+use tokio::{process, runtime, sync};
+
+use crate::cli_util::{CommandHelper, RevisionArg, WorkspaceCommandTransaction};
 use crate::command_error::{user_error, CommandError};
 use crate::ui::Ui;
 
+#[derive(Debug, thiserror::Error)]
+enum RunError {
+    #[error("Couldn't create directory")]
+    NoDirectoryCreated,
+    #[error("the command failed {} for {}")]
+    CommandFailure(ExitCode, CommitId),
+}
+
+impl From<RunError> for CommandError {
+    fn from(value: RunError) -> Self {
+        CommandError::new(crate::command_error::CommandErrorKind::Cli, Box::new(value))
+    }
+}
+
+/// Creates the required directories for a StoredWorkingCopy.
+/// Returns a tuple of (`output_dir`, `working_copy` and `state`).
+fn create_working_copy_paths(
+    path: &PathBuf,
+) -> Result<(PathBuf, PathBuf, PathBuf), std::io::Error> {
+    let output = path.join("output");
+    let working_copy = path.join("working_copy");
+    let state = path.join("state");
+    std::fs::create_dir(&output)?;
+    std::fs::create_dir(&working_copy)?;
+    std::fs::create_dir(&state)?;
+    Ok((output, working_copy, state))
+}
+
+/// Represent a `MergeTreeId` in a way that it may be used as a working-copy
+/// name. This makes no stability guarantee, as the format may change at
+/// any time.
+fn to_wc_name(id: &MergedTreeId) -> String {
+    match id {
+        MergedTreeId::Legacy(tree_id) => tree_id.hex(),
+        MergedTreeId::Merge(tree_ids) => {
+            let ids = tree_ids
+                .map(|id| id.hex())
+                .iter_mut()
+                .enumerate()
+                .map(|(i, s)| {
+                    // Incredibly "smart" way to say, append "-" if the number is odd "+"
+                    // otherwise.
+                    if i & 1 != 0 {
+                        s.push('-');
+                    } else {
+                        s.push('+');
+                    }
+                    s.to_owned()
+                })
+                .collect_vec();
+            let mut obfuscated: String = ids.concat();
+            // `PATH_MAX` could be a problem for different operating systems, so truncate
+            // it.
+            if obfuscated.len() >= 255 {
+                obfuscated.truncate(200);
+            }
+            obfuscated
+        }
+    }
+}
+
+fn get_runtime(jobs: usize) -> tokio::runtime::Handle {
+    let mut builder = Builder::new_multi_thread();
+    if cfg!(watchman) {
+        // Watchman requires a multithreaded runtime, so just reuse it.
+        return runtime::Handle::current();
+    }
+    if jobs == 1 {
+        builder.max_blocking_threads(1);
+    } else {
+        builder.max_blocking_threads(jobs);
+    }
+    let rt = builder.build().unwrap();
+    rt.handle().clone()
+}
+
+/// A commit stored under `.jj/run/`
+// TODO: Create a caching backend, which creates these on a dedicated thread or
+// threadpool.
+struct StoredCommit {
+    /// Obfuscated name for an easier lookup. If a tree/directory its not set
```

Comment:
question: Easier lookup by whom, for what purpose?

Reply:
We need to replace the commits stored under `.jj/run/default/` if we have commits > jobs and that in-place. So the canonical commit is the initially created one, while those who overwrite it won't have a name. If that makes sense.